test/e2e/autoscale_test.go

```diff
@@ -339,8 +348,95 @@ func TestAutoscaleSustaining(t *testing.T) {
 	cancel := logstream.Start(t)
 	defer cancel()
 
-	ctx := setup(t, autoscaling.KPA, autoscaling.Concurrency)
+	ctx := setup(t, autoscaling.KPA, autoscaling.Concurrency, &v1a1test.Options{
+		ContainerConcurrency: containerConcurrency,
+	})
 	defer test.TearDown(ctx.clients, ctx.names)
 
 	assertAutoscaleUpToNumPods(ctx, 1, 10, 3*time.Minute, false)
 }
+
+func TestTargetBurstCapacity(t *testing.T) {
+	// This test sets up a service with CC=10 TU=70% and TBC=7.
+	// Then sends requests at concurrency 5, causing Activator in the path (1 pod, 10-5=5 < 7).
+	// Then at the concurrency 10, getting spare capacity of 20-10=10, which should remove the
+	// Activator from the request path.
+	cancel := logstream.Start(t)
+	defer cancel()
+
+	ctx := setup(t, autoscaling.KPA, autoscaling.Concurrency, &v1a1test.Options{},
+		rtesting.WithConfigAnnotations(map[string]string{
+			autoscaling.TargetAnnotationKey:                   "10",
+			autoscaling.TargetUtilizationPercentageKey:        "70",
+			autoscaling.TargetBurstCapacityKey:                "7",
+			autoscaling.PanicThresholdPercentageAnnotationKey: "200", // makes panic'ing rare
+		}))
+	defer test.TearDown(ctx.clients, ctx.names)
+
+	cfg, err := autoscalerCM(ctx.clients)
+	if err != nil {
+		t.Fatalf("Error retrieving autoscaler configmap: %v", err)
+	}
+	var (
+		grp    errgroup.Group
+		stopCh = make(chan struct{})
+	)
+	defer grp.Wait()
+	defer close(stopCh)
+
+	// We'll terminate the test via stopCh.
+	const duration = time.Hour
+
+	grp.Go(func() error {
+		return generateTraffic(ctx, 5, duration, stopCh)
+	})
+	aeps, err := ctx.clients.KubeClient.Kube.CoreV1().Endpoints(system.Namespace()).Get(activator.K8sServiceName, metav1.GetOptions{})
+	if err != nil {
+		t.Fatalf("Error getting activator endpoints: %v", err)
+	}
+	t.Logf("Activator endpoints: %v", aeps)
+
+	// Wait for the endpoints to equalize.
+	// There's some jitter in the beginning possible, since it takes 2 seconds
+	// for us to start evaluating things.
+	if err := wait.Poll(250*time.Millisecond, cfg.StableWindow, func() (bool, error) {
+		svcEps, err := ctx.clients.KubeClient.Kube.CoreV1().Endpoints(test.ServingNamespace).Get(
+			ctx.resources.Revision.Status.ServiceName, metav1.GetOptions{})
+		if err != nil {
+			return false, err
+		}
+		return cmp.Equal(svcEps.Subsets, aeps.Subsets), nil
+	}); err != nil {
+		t.Fatalf("Initial state never achieved: %v", err)
+	}
+
+	// Start second load generator.
+	grp.Go(func() error {
+		return generateTraffic(ctx, 5, duration, stopCh)
+	})
+
+	// Wait for two stable pods.
+	if err := wait.Poll(250*time.Millisecond, cfg.StableWindow, func() (bool, error) {
+		x, err := numberOfPods(ctx)
```

Comment:
Is this check redundant with the 2 endpoints check? After you see two endpoints, it should be safe to check for 2 pods without polling.

Reply:
Well, I went about this the following way:

1.  K8s creates the pod
2. when it's healthy it's added to the service

So, if I test for the pod creation first then I know that precondition for 2 is satisfied.

Thus if there are problems with the deployment that created the pod but endpoint doesn't show up, it's easier to narrow down (health check, SKS bug)